### PR TITLE
Torch support [DON'T MERGE]

### PR DESCRIPTION
**Pull request by lukeyeager**
Thanks very much to @Sravan2j for his hard work on this! This is going to be a somewhat long-lived feature branch.

* Please open pull requests for torch-related changes against this branch, not against `master`.
* Since this is now a public branch, **don't do any rebasing**. Just merge in changes from `master` to `torch` periodically. There may be a big rebase at the last minute before merging into master.
* Where possible, try to open pull requests against `master` for changes that are generic across frameworks, like `task.framework_name()`. Let's try to break some of this up a bit.

Comments/critiques welcome!

### CODE REVIEW

**Comment by lukeyeager** [comment]
@Alexis-Jacob, any comments on this? How does this line up with the work you guys have been doing at @Deepomatic?

**Comment by Alexis-Jacob** [comment]
First of all, everything that you've made is great and covers most of what we did and were planning to do at @deepomatic

However I have some questions/suggestions: 
* Is there a way to launch digits without caffe ? I believe that digits should be framework independent
* During the creation of the dataset:
    * if torch is installed maybe we should use torch instead of caffe to generate the db (in order to use specific torch features such as [the color space conversions](https://github.com/torch/image#color-space-conversions) for instance)

**Comment by dzhwinter** [comment]
Adds, It should be framework independent .It should be launch without caffe or any other thrid party framework

**Comment by Sravan2j** [comment]
@Alexis-Jacob and @dzhwinter,  yes, that makes sense to launch digits without caffe. I will check what all changes are required to implement that. Also regarding the usage of torch specific feature (i.e., the color space conversions), there is a resize function "DIGITS/tools/torch/utils.lua" module, probably we may extend that. What are your suggestions? Please let me know

File: digits/model/forms.py [hunk 1]
```diff
@@ -154,6 +155,17 @@ def validate_lr_multistep_values(form, field):
             default = 'standard',
             )
 
+    ## framework
+    framework = wtforms.HiddenField('framework',
+            validators = [
+                validators.AnyOf(
+                    ['caffe', 'torch', 'theano'],
```
Comment (nsubtil) [inline digits/model/forms.py]: Is theano supported? Should we check for theano here?

File: digits/model/images/classification/views.py [hunk 2]
```diff
@@ -236,12 +289,10 @@ def image_classification_model_classify_many():
     paths = []
     images = []
     dataset = job.train_task().dataset
-
```
Comment (nsubtil) [inline digits/model/images/classification/views.py]: Why the whitespace changes?

File: digits/model/forms.py (hunk 1 above)
Comment (Sravan2j) [inline digits/model/forms.py]: No Theano isn't supported. I will remove that. Thanks :)

**Comment by lukeyeager** [comment]
I tried pulling this code and ran into several issues immediately.

### Error messages on job submission

I just tried running MNIST, and DIGITS gave me this error:

    useMeanPixel parameter is not considered as subtractMean value is provided as "yes"
    Loading Model: /raid/jobs/dev/20150518-164856-686f/model

1. Is supplying both `useMeanPixel` and `subtractMean` an error? Sounds like it should be a warning at most. And why did DIGITS try to use both flags anyway?
2. Why is torch attempting to "load a model"? I didn't use a pretrained model.

### Checking for unmet dependencies

Here is the `torch_output.log` from that job. It looks like I'm missing some dependencies.

```
2015-05-18 16:49:00 [INFO ] useMeanPixel parameter is not considered as subtractMean value is provided as "yes"
2015-05-18 16:49:00 [INFO ] Loading Model: /raid/jobs/dev/20150518-164856-686f/model
/home/lyeager/torch/install/bin/luajit: /home/lyeager/torch/install/share/lua/5.1/trepl/init.lua:319: /home/lyeager/torch/install/share/lua/5.1/trepl/init.lua:319: module 'inn' not found:No LuaRocks module found for inn
no field package.preload['inn']
no file '/raid/jobs/dev/20150518-164856-686f/inn.lua'
no file '/home/lyeager/digits/tools/torch/inn.lua'
no file '/home/lyeager/.luarocks/share/lua/5.1/inn.lua'
no file '/home/lyeager/.luarocks/share/lua/5.1/inn/init.lua'
no file '/home/lyeager/torch/install/share/lua/5.1/inn.lua'
no file '/home/lyeager/torch/install/share/lua/5.1/inn/init.lua'
no file './inn.lua'
no file '/home/lyeager/torch/install/share/luajit-2.1.0-alpha/inn.lua'
no file '/usr/local/share/lua/5.1/inn.lua'
no file '/usr/local/share/lua/5.1/inn/init.lua'
no file '/home/lyeager/.luarocks/lib/lua/5.1/inn.so'
no file '/home/lyeager/torch/install/lib/lua/5.1/inn.so'
no file './inn.so'
no file '/usr/local/lib/lua/5.1/inn.so'
no file '/usr/local/lib/lua/5.1/loadall.so'
stack traceback:
[C]: in function 'error'
/home/lyeager/torch/install/share/lua/5.1/trepl/init.lua:319: in function 'require'
/home/lyeager/digits/tools/torch/main.lua:156: in main chunk
[C]: in function 'dofile'
...ager/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
[C]: at 0x00406170
```
That's probably my fault for not following the Torch install instructions. But is there any way to detect whether the user has installed all the necessary requirements? I do this for caffe in the `apply()` function [in the config module](https://github.com/NVIDIA/DIGITS/blob/0089c408605d892d0426389851ed6fbb0178cbc0/digits/config.py#L418-L422).

### Viewing caffe jobs

When I try to view the page for a job that was created with caffe, I get this error:

    BuildError
    ('serve_file', {}, None)

By running DIGITS with the `--debug` switch, I see that the problem is [on this line](https://github.com/NVIDIA/DIGITS/blob/f5d350cae19b5cda0167eac17652742c7ca0d0e9/digits/templates/models/images/classification/show.html#L30). I think the problem is that `task.log_file` isn't set. What'd you do there?

File: docs/TorchInstall.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+export LUA_CPATH="&lt;LIGHTININGMDB_INSTALLATION_DIR&gt;//?.so;;"
+
+For example: 
+export LUA_PATH="/home/ubuntu/Downloads/lightningdbm//?.lua;;"
+export LUA_CPATH="/home/ubuntu/Downloads/lightningdbm//?.so;;"
+</pre>
+
+Instead of setting the paths in `~/.profile` file, we can also place the `lightningmdb.so` file in `<TORCH_INSTALLATION_DIR>/lib` directory
+
+Please check this link for more information regarding lightningdbm: `https://github.com/shmul/lightningdbm`
```
Comment (lukeyeager) [inline docs/TorchInstall.md]: Can we just tell people how to install with `luarocks` instead of asking them to edit the Makefile manually?

**Comment by lukeyeager** [comment]
Ok, now I think I've installed all the requirements (following your instructions), and I get this:

```
2015-05-18 18:29:08 [INFO ] useMeanPixel parameter is not considered as subtractMean value is provided as "yes"
2015-05-18 18:29:08 [INFO ] Loading Model: /raid/jobs/dev/20150518-182845-834a/model
/home/lyeager/torch/install/bin/luajit: /home/lyeager/torch/install/share/lua/5.1/trepl/init.lua:363: /home/lyeager/torch/install/share/lua/5.1/trepl/init.lua:363: /home/lyeager/torch/install/share/lua/5.1/pb.lua:212:
no file '/raid/jobs/dev/20150518-182845-834a/lightningmdb.proto'
no file '/home/lyeager/digits/tools/torch/lightningmdb.proto'
no file '/home/lyeager/.luarocks/share/lua/5.1/lightningmdb.proto'
no file '/home/lyeager/torch/install/share/lua/5.1/lightningmdb.proto'
no file './lightningmdb.proto'
no file '/home/lyeager/torch/install/share/luajit-2.1.0-alpha/lightningmdb.proto'
no file '/usr/local/share/lua/5.1/lightningmdb.proto'
stack traceback:
[C]: in function 'error'
/home/lyeager/torch/install/share/lua/5.1/trepl/init.lua:363: in function 'require'
/home/lyeager/digits/tools/torch/main.lua:169: in main chunk
[C]: in function 'dofile'
...ager/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
[C]: at 0x00406670
```

I can't grok what's gone wrong here. Did I miss some package I needed?

**Comment by lukeyeager** [comment]
Why doesn't the torch MNIST network work on color images? Is that something that can be fixed?
```
2015-05-27 10:14:17 [INFO ] started training the model
/home/lyeager/torch/install/bin/luajit: ...er/torch/install/share/lua/5.1/nn/SpatialConvolution.lua:88: bad argument #2 to 'SpatialConvolutionMM_updateOutput' (input channels and nInputPlane dont match)
stack traceback:
[C]: in function 'SpatialConvolutionMM_updateOutput'
...er/torch/install/share/lua/5.1/nn/SpatialConvolution.lua:88: in function 'updateOutput'
/home/lyeager/torch/install/share/lua/5.1/nn/Sequential.lua:29: in function 'forward'
/home/lyeager/digits/tools/torch/main.lua:526: in function 'Test'
/home/lyeager/digits/tools/torch/main.lua:665: in main chunk
[C]: in function 'dofile'
...ager/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
[C]: at 0x00406670
```

**Comment by lukeyeager** [comment]
Closing in favor of #324.

**Pull request closed without merging**